src/compiler/scanner.ts

```diff
@@ -1882,69 +1996,115 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
+                if (skipTrivia) {
+                    pos++;
+                    continue;
+                }
+                else {
+                    if (ch === CharacterCodes.carriageReturn && pos + 1 < end && charCodeUnchecked(pos + 1) === CharacterCodes.lineFeed) {
+                        // consume both CR and LF
+                        pos += 2;
                     }
                     else {
-                        return token = SyntaxKind.ShebangTrivia;
+                        pos++;
                     }
+                    return token = SyntaxKind.NewLineTrivia;
                 }
             }
 
-            switch (ch) {
-                case CharacterCodes.lineFeed:
-                case CharacterCodes.carriageReturn:
-                    tokenFlags |= TokenFlags.PrecedingLineBreak;
-                    if (skipTrivia) {
+            const tokenCategory = ch < tokenCategoryLookup.length ?
+                tokenCategoryLookup[ch] :
+                tokenCategoryLookupUncommon.get(ch) ?? TokenCategory.IdentifierOrUnknown;
+
+            if (tokenCategory === TokenCategory.IdentifierOrUnknown) {
+                const identifierKind = scanIdentifier(ch, languageVersion);
+                if (identifierKind) {
+                    return token = identifierKind;
+                }
+                const size = charSize(ch);
+                error(Diagnostics.Invalid_character, pos, size);
+                pos += size;
+                return token = SyntaxKind.Unknown;
+            }
+
+            if (tokenCategory & TokenCategory.Whitespace) {
+                pos++;
+                // Tight loop here on consecutive whitespace to avoid a table lookup.
+                while (pos < end) {
+                    const nextCh = charCodeUnchecked(pos);
+                    // Check for the original character to hitting the slow path.
+                    if (nextCh === ch || isWhiteSpaceSingleLine(nextCh)) {
                         pos++;
                         continue;
                     }
-                    else {
-                        if (ch === CharacterCodes.carriageReturn && pos + 1 < end && charCodeUnchecked(pos + 1) === CharacterCodes.lineFeed) {
-                            // consume both CR and LF
-                            pos += 2;
+
+                    break;
+                }
+
+                if (skipTrivia) {
+                    continue;
+                }
+                else {
+                    return token = SyntaxKind.WhitespaceTrivia;
+                }
+            }
+
+            if (tokenCategory & TokenCategory.LineBreak) {
+                tokenFlags |= TokenFlags.PrecedingLineBreak;
+                pos += charSize(ch);
+                continue;
+            }
+
+            if (tokenCategory & TokenCategory.SimpleToken) {
+                pos++;
+                return token = tokenCategory & TokenCategory.SimpleTokenMask;

```

Comment:
A comment here would be good for future readers as it isn't obvious at first glance how a `TokenCategory` becomes a token.